# Anime Recommendation Engine

As per Wikipedia, Anime is hand-drawn and computer animation originating from Japan. The word anime is the Japanese term for animation, which refers to all forms of animated media. Nowadays, a lot of people are finding anime as a great form of entertainment. With its distinctive artstyle and voice-acting, it has garnered a lot of appreciation especially among the youths. As an anime fan myself, I would be exploring and also I will be building an anime recommendation engine based on the dataset - Anime Recommendation Database - which was scrapped from myanimelist.net (also known as MAL)  - a website which facilitates finding users who share similar tastes and provides a large database on anime and manga.

## Data

The dataset we will be using for building the recommendation engine will be the [Anime Recommendation Database](https://www.kaggle.com/CooperUnion/anime-recommendations-database) which has two files - anime.csv and rating.csv. This data set contains information on user preference data from 73,516 users on 12,294 anime. Each user is able to add anime to their completed list and give it a rating and this data set is a compilation of those ratings.

**anime.csv**
1. anime_id - myanimelist.net's unique id identifying an anime.
2. name - full name of anime.
3. genre - comma separated list of genres for this anime.
4. type - movie, TV, OVA, etc.
5. episodes - how many episodes in this show. (1 if movie).
6. rating - average rating out of 10 for this anime.
7. members - number of community members that are in this anime's "group".

**rating.csv**
1. user_id - non identifiable randomly generated user id.
2. anime_id - the anime that this user has rated.
3. rating - rating out of 10 this user has assigned (-1 if the user watched it but didn't assign a rating).

<h2>Part 1 Data Prepration and Cleaning</h2>
    <p>Objective</p>
    <ol><li>Import all required libraries</li>
    <li>Find initial dimension of data</li>
        <li>Print the first and the last 5 rows.</li>
        <li>Convert the features to their appropriate data types </li>
        <li>Drop duplicate rows and columns, if any.</li>

<h3>Import libraries</h3>
<p> The most important library used is the <b>pandas</b>, this enables the csv to be converted to a pandas datatype and allow data manipulation. <b>Numpy</b> library adds array like functionalities to python for scientific calculations. The visulaization is carried out using <b>Matplotlib</b> and <b>Seaborn</b>.

In [6]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from IPython.display import clear_output
import seaborn as sns

In [7]:
# This option enables auto complete in notebook
%config IPCompleter.greedy=True

### Loading the two csv files

In [5]:
# Load csv to pandas dataframe
anime = pd.read_csv('anime.csv')
# df_test is the test dataset which will also be pre-processed together
rating = pd.read_csv('rating.csv')
# Enables viewing of all columns when using the head and tail command
pd.options.display.max_columns = None
pd.options.display.max_rows = None


# Show full column with out truncation
pd.set_option('display.max_colwidth', -1)

### Find the initial dimension of dataframe

In [9]:
print ('anime shape: ', anime.shape)
print ('rating shape: ', rating.shape)

anime shape:  (12294, 7)
rating shape:  (7813737, 3)


<blockquote>The dataframe anime initially has a dimension of 12294 x 7 which means there are 7 columns of attributes and 12294 rows of data. The dataframe rating initially has a dimension of 7813737 x 3 which means there are 3 columns of attributes and 7813737 rows of data.</blockquote>

### Getting an overview of data
by checking first and last 5 rows of the dataframes

In [10]:
# Checking first 5 rows of anime dataframe
anime.head(5)

anime_id                              name  \
0  32281     Kimi no Na wa.                     
1  5114      Fullmetal Alchemist: Brotherhood   
2  28977     Gintama°                           
3  9253      Steins;Gate                        
4  9969      Gintama&#039;                      

                                                          genre   type  \
0  Drama, Romance, School, Supernatural                          Movie   
1  Action, Adventure, Drama, Fantasy, Magic, Military, Shounen   TV      
2  Action, Comedy, Historical, Parody, Samurai, Sci-Fi, Shounen  TV      
3  Sci-Fi, Thriller                                              TV      
4  Action, Comedy, Historical, Parody, Samurai, Sci-Fi, Shounen  TV      

  episodes  rating  members  
0  1        9.37    200630   
1  64       9.26    793665   
2  51       9.25    114262   
3  24       9.17    673572   
4  51       9.16    151266

In [12]:
# Checking the last 5 rows of anime dataframe
anime.tail(5)

anime_id                                                name   genre  \
12289  9316      Toushindai My Lover: Minami tai Mecha-Minami        Hentai   
12290  5543      Under World                                         Hentai   
12291  5621      Violence Gekiga David no Hoshi                      Hentai   
12292  6133      Violence Gekiga Shin David no Hoshi: Inma Densetsu  Hentai   
12293  26081     Yasuji no Pornorama: Yacchimae!!                    Hentai   

        type episodes  rating  members  
12289  OVA    1        4.15    211      
12290  OVA    1        4.28    183      
12291  OVA    4        4.88    219      
12292  OVA    1        4.98    175      
12293  Movie  1        5.46    142

In [13]:
# Checking the first 5 rows of rating dataframe
rating.head(5)

user_id  anime_id  rating
0  1        20       -1     
1  1        24       -1     
2  1        79       -1     
3  1        226      -1     
4  1        241      -1

In [14]:
# Checking the last 5 rows of rating dataframe
rating.tail(5)

user_id  anime_id  rating
7813732  73515    16512     7     
7813733  73515    17187     9     
7813734  73515    22145     10    
7813735  73516    790       9     
7813736  73516    8074      9

### Checking datatypes of all attributes

In [15]:
# Checking for anime dataframe
anime.dtypes

anime_id    int64  
name        object 
genre       object 
type        object 
episodes    object 
rating      float64
members     int64  
dtype: object

In [16]:
# Checking for rating dataframe
rating.dtypes

user_id     int64
anime_id    int64
rating      int64
dtype: object

In [17]:
anime.describe().T

count          mean           std   min      25%       50%  \
anime_id  12294.0  14058.221653  11455.294701  1.00  3484.25  10260.50   
rating    12064.0  6.473902      1.026746      1.67  5.88     6.57       
members   12294.0  18071.338864  54820.676925  5.00  225.00   1550.00    

               75%        max  
anime_id  24794.50  34527.0    
rating    7.18      10.0       
members   9437.00   1013917.0

In [18]:
rating.describe().T

count          mean           std  min      25%      50%  \
user_id   7813737.0  36727.956745  20997.946119  1.0  18974.0  36791.0   
anime_id  7813737.0  8909.072104   8883.949636   1.0  1240.0   6213.0    
rating    7813737.0  6.144030      3.727800     -1.0  6.0      7.0       

              75%      max  
user_id   54757.0  73516.0  
anime_id  14093.0  34519.0  
rating    9.0      10.0

### Finding out all the null values in the attributes

In [22]:
# For anime dataframe
anime[anime.columns].isnull().sum()

anime_id    0  
name        0  
genre       62 
type        25 
episodes    0  
rating      230
members     0  
dtype: int64

In [23]:
# For rating dataframe
rating[rating.columns].isnull().sum()

user_id     0
anime_id    0
rating      0
dtype: int64

<blockquote>From the above information we can see that there are 62 missing values in genre feature, 25 in feature type as well as 230 in the rating feature. Since we are building a recommendation engine using collabarative filtering we will ony look at the relative rating of animes by each users.</blockquote>


## Exploratory Data Analysis and Visualisation
<b> We will be exploring the two dataframes, anime and rating so as to analyse and gain more insights from the dataset.</b> 

In [40]:
def get_cardinalities(columns):
    """To get the unique cardinalities of a column in a dataframe"""
    colist = columns.tolist()
    card = []
    for i in colist:
        k = list(str(i).split(', '))
        for j in k:
            if j not in card:
                card.append(j)
    return card

In [50]:
genre_ = get_cardinalities(anime.genre)
type_ = get_cardinalities(anime.type)
print ('There are over {} unique entries to genre feature in the anime dataframe.'.format(len(genre_)))
print ('There are over {} unique entries to type feature in the anime dataframe.'.format(len(type_)))

There are over 44 unique entries to genre feature in the anime dataframe.
There are over 7 unique entries to genre feature in the anime dataframe.


In [68]:
user = rating.user_id.unique()
print ('There are {} number of unique user ids in the rating dataframe.'.format(len(user)))

There are 73515 number of unique user ids in rating dataframe
